drivers/gpio/gpio_sifive.c

```diff
@@ -284,6 +269,110 @@ static int gpio_sifive_read(struct device *dev,
 	return 0;
 }
 
+static int gpio_sifive_port_get_raw(struct device *dev,
+				   gpio_port_value_t *value)
+{
+	volatile struct gpio_sifive_t *gpio = DEV_GPIO(dev);
+
+	*value = gpio->in_val;
+
+	return 0;
+}
+
+static int gpio_sifive_port_set_masked_raw(struct device *dev,
+					  gpio_port_pins_t mask,
+					  gpio_port_value_t value)
+{
+	volatile struct gpio_sifive_t *gpio = DEV_GPIO(dev);
+
+	gpio->out_val = (gpio->out_val & ~mask) | (value & mask);
+
+	return 0;
+}
+
+static int gpio_sifive_port_set_bits_raw(struct device *dev,
+					gpio_port_pins_t mask)
+{
+	volatile struct gpio_sifive_t *gpio = DEV_GPIO(dev);
+
+	gpio->out_val |= mask;
+
+	return 0;
+}
+
+static int gpio_sifive_port_clear_bits_raw(struct device *dev,
+					  gpio_port_pins_t mask)
+{
+	volatile struct gpio_sifive_t *gpio = DEV_GPIO(dev);
+
+	gpio->out_val &= ~mask;
+
+	return 0;
+}
+
+static int gpio_sifive_port_toggle_bits(struct device *dev,
+				       gpio_port_pins_t mask)
+{
+	volatile struct gpio_sifive_t *gpio = DEV_GPIO(dev);
+
+	gpio->out_val ^= mask;
+
+	return 0;
+}
+
+static int gpio_sifive_pin_interrupt_configure(struct device *dev,
+					      unsigned int pin,
+					      enum gpio_int_mode mode,
+					      enum gpio_int_trig trig)
+{
+	volatile struct gpio_sifive_t *gpio = DEV_GPIO(dev);
+	const struct gpio_sifive_config *cfg = DEV_GPIO_CFG(dev);
+
+	switch (mode) {
+	case GPIO_INT_MODE_DISABLED:
+		gpio->rise_ie &= ~BIT(pin);
+		gpio->rise_ie &= ~BIT(pin);
+		gpio->fall_ie &= ~BIT(pin);
+		gpio->high_ie &= ~BIT(pin);
+		gpio->low_ie &= ~BIT(pin);
+		irq_disable(gpio_sifive_pin_irq(cfg->gpio_irq_base, pin));
+		break;
+	case GPIO_INT_MODE_LEVEL:
+		gpio->rise_ie &= ~BIT(pin);
+		gpio->rise_ie &= ~BIT(pin);
+		gpio->fall_ie &= ~BIT(pin);
+
+		if (trig == GPIO_INT_TRIG_HIGH) {
+			gpio->high_ie |= BIT(pin);
+			gpio->low_ie &= ~BIT(pin);
+		} else if (trig == GPIO_INT_TRIG_LOW) {
+			gpio->high_ie &= ~BIT(pin);
+			gpio->low_ie |= BIT(pin);
+		}
+		irq_enable(gpio_sifive_pin_irq(cfg->gpio_irq_base, pin));
```

Comment:
I don't know this platform but we probably need to clear interrupt status register before enabling the interrupt source. E.g. in the irq handler we have the following code
```
	if (gpio->rise_ip & BIT(pin)) {
		gpio->rise_ip = BIT(pin);
	} else if (gpio->fall_ip & BIT(pin)) {
		gpio->fall_ip = BIT(pin);
	} else if (gpio->high_ip & BIT(pin)) {
		gpio->high_ip = BIT(pin);
	} else if (gpio->low_ip & BIT(pin)) {
		gpio->low_ip = BIT(pin);
	}
```
On most of the hardware having 'dirty' interrupt status register would result in an interrupt being triggered immediately.